pkg/storage/fs/local/upload.go

```diff
@@ -0,0 +1,290 @@
+	return filepath.Join(fs.conf.Uploads, uploadID), nil
+}
+
+// GetUpload returns the Upload for the given upload id
+func (fs *localfs) GetUpload(ctx context.Context, id string) (tusd.Upload, error) {
+	binPath, err := fs.getUploadPath(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	infoPath := binPath + ".info"
+	info := tusd.FileInfo{}
+	data, err := ioutil.ReadFile(infoPath)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(data, &info); err != nil {
+		return nil, err
+	}
+
+	stat, err := os.Stat(binPath)
+	if err != nil {
+		return nil, err
+	}
+
+	info.Offset = stat.Size()
+
+	return &fileUpload{
+		info:     info,
+		binPath:  binPath,
+		infoPath: infoPath,
+		fs:       fs,
+	}, nil
+}
+
+type fileUpload struct {
+	// info stores the current information about the upload
+	info tusd.FileInfo
+	// infoPath is the path to the .info file
+	infoPath string
+	// binPath is the path to the binary file (which has no extension)
+	binPath string
+	// only fs knows how to handle metadata and versions
+	fs *localfs
+}
+
+// GetInfo returns the FileInfo
+func (upload *fileUpload) GetInfo(ctx context.Context) (tusd.FileInfo, error) {
+	return upload.info, nil
+}
+
+// GetReader returns an io.Readerfor the upload
+func (upload *fileUpload) GetReader(ctx context.Context) (io.Reader, error) {
+	return os.Open(upload.binPath)
+}
+
+// WriteChunk writes the stream from the reader to the given offset of the upload
+func (upload *fileUpload) WriteChunk(ctx context.Context, offset int64, src io.Reader) (int64, error) {
+	file, err := os.OpenFile(upload.binPath, os.O_WRONLY|os.O_APPEND, defaultFilePerm)
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	n, err := io.Copy(file, src)
+
+	// If the HTTP PATCH request gets interrupted in the middle (e.g. because
+	// the user wants to pause the upload), Go's net/http returns an io.ErrUnexpectedEOF.
+	// However, for OwnCloudStore it's not important whether the stream has ended
+	// on purpose or accidentally.
+	if err != nil {
+		if err != io.ErrUnexpectedEOF {
+			return n, err
+		}
+	}
+
+	upload.info.Offset += n
+	err = upload.writeInfo()
+
+	return n, err
+}
+
+// writeInfo updates the entire information. Everything will be overwritten.
+func (upload *fileUpload) writeInfo() error {
+	data, err := json.Marshal(upload.info)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(upload.infoPath, data, defaultFilePerm)
+}
+
+// FinishUpload finishes an upload and moves the file to the internal destination
+func (upload *fileUpload) FinishUpload(ctx context.Context) error {
+
+	np := upload.info.Storage["InternalDestination"]
+
+	// TODO check etag with If-Match header
+	// if destination exists
+	//if _, err := os.Stat(np); err == nil {
+	// the local storage does not store metadata
+	// the fileid is based on the path, so no we do not need to copy it to the new file
+	// the local storage does not track revisions
+	//}
+
+	err := os.Rename(upload.binPath, np)
```

Comment:
we might also need a way to return the etag and oc-fileid for desktop client and others, like for the older PUT requests

cc @TheOneRing

Reply:
This is the tus upload implementaion. If we need to return the etag that would have to be done in the tus handler. Well, or using the hook mechanism ... for ocdav, our webdav api, we can execute a stat request after the upload via tus.

But if the clients are speaking tus directly, indeed, we should be able to return arbitrary metadata ... 

Hm. I think this should not be part of tus, because the bytes have been transferred. the file metadata is now available in the cs3 api, which leads to another question: How long does it take for the file to become available for download? There might be antivirus scanning going on etc. Tracked in https://github.com/owncloud/ocis/issues/214

As long as this is hidden behind ocdav the stat fallback works ... transloadid uses websockets to send progress notifications of the 'assembly' to the clients. It would allow at least phoenix to show a progresspar that includes detailed postprocessing steps.